### PR TITLE
Add spirv-cfg to dump a GraphViz graph of the CFG

### PR DESCRIPTION
**Pull request by dneto0**
This is experimental, and has not tests.
It's been used to debug validation of structured control flow.

### CODE REVIEW

**Comment by dneto0** [comment]
Rebased

File: tools/cfg/cfg.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+#include <string>
+#include <vector>
+
+#include "spirv-tools/libspirv.h"
+#include "tools/io.h"
+
+#include "bin_to_dot.h"
+
+// Prints a program usage message to stdout.
+static void print_usage(char* argv0) {
```
Comment (antiagainst) [inline tools/cfg/cfg.cpp]: `const char*`?

File: tools/cfg/cfg.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  std::vector<uint32_t> contents;
+  if (!ReadFile<uint32_t>(inFile, "rb", &contents)) return 1;
+  spv_context context = spvContextCreate(SPV_ENV_UNIVERSAL_1_1);
+  spv_diagnostic diagnostic = nullptr;
+
+  std::stringstream ss;
+  auto error = BinaryToDot(context, contents.data(), contents.size(), &ss, &diagnostic);
+  if (error) {
+    spvDiagnosticPrint(diagnostic);
+    spvDiagnosticDestroy(diagnostic);
```
Comment (antiagainst) [inline tools/cfg/cfg.cpp]: need `spvContextDestroy` here.

File: tools/cfg/cfg.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    return error;
+  }
+  std::string str = ss.str();
+  WriteFile(outFile, "w", str.data(), str.size());
+
```
Comment (antiagainst) [inline tools/cfg/cfg.cpp]: need `spvDiagnosticDestroy` here.

File: tools/cfg/bin_to_dot.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+#ifndef BIN_TO_DOT_H_
+#define BIN_TO_DOT_H_
+
+#include <iostream>
+#include "spirv-tools/libspirv.h"
+
+// Dumps the control flow graph for the given module to the output stream.
+// Returns SPV_SUCCESS on succes.
+spv_result_t BinaryToDot(const spv_const_context context, uint32_t* words,
```
Comment (antiagainst) [inline tools/cfg/bin_to_dot.h]: `const uint32_t*`?

File: tools/cfg/bin_to_dot.cpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+  // The ID of the current basic block, or 0 if outside of a block.
+  uint32_t current_block_id_ = 0;
+
+  // Have we completed processing for the entry block to this fuction?
+  bool seen_function_entry_block_ = false;
+
+  // The Id of the merge block for this block if it exists, or 0 otherwise.
+  uint32_t merge_ = 0;
+  // The Id of the continue block for this block if it exists, or 0 otherwise.
```
Comment (antiagainst) [inline tools/cfg/bin_to_dot.cpp]: Continue block or continue target block?

My understanding is they are different. And quite easily get confused :)

File: tools/cfg/bin_to_dot.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+      break;
+    case SpvOpSelectionMerge:
+      merge_ = inst.words[1];
+      break;
+    default:
+      break;
+  }
+  return SPV_SUCCESS;
+}
+
```
Comment (antiagainst) [inline tools/cfg/bin_to_dot.cpp]: It would be nice to add a comment to explain those `shape`s and `style`s in the output.

**Comment by antiagainst** [comment]
Logic LGTM, just some nits.

File: tools/cfg/bin_to_dot.cpp (hunk 5 above)
Comment (dneto0) [inline tools/cfg/bin_to_dot.cpp]: Good catch.  Continue target bock.

**Comment by dneto0** [comment]
Rebased, and responded to feedback in 75b4477

**Comment by dneto0** [comment]
Sample output: 
vert

**Comment by antiagainst** [comment]
+2

**Comment by dneto0** [comment]
Rebased.
+2

**Pull request merged**